// @eslint/eslintrc/lib/config-array/config-dependency.js
 

import util from "util";

 
class ConfigDependency {

     
    constructor({
        definition = null,
        original = null,
        error = null,
        filePath = null,
        id,
        importerName,
        importerPath
    }) {

         
        this.definition = definition;

         
        this.original = original;

         
        this.error = error;

         
        this.filePath = filePath;

         
        this.id = id;

         
        this.importerName = importerName;

         
        this.importerPath = importerPath;
    }

    // eslint-disable-next-line jsdoc/require-description
     
    toJSON() {
        const obj = this[util.inspect.custom]();

        // Display `error.message` (`Error#message` is unenumerable).
        if (obj.error instanceof Error) {
            obj.error = { ...obj.error, message: obj.error.message };
        }

        return obj;
    }

    // eslint-disable-next-line jsdoc/require-description
     
    [util.inspect.custom]() {
        const {
            definition: _ignore1, // eslint-disable-line no-unused-vars
            original: _ignore2, // eslint-disable-line no-unused-vars
            ...obj
        } = this;

        return obj;
    }
}

 
 

export { ConfigDependency };
