// @eslint/eslintrc/lib/config-array/extracted-config.js
 

import { IgnorePattern } from "./ignore-pattern.js";

// For VSCode intellisense
 
 
 
 
 

 
function startsWith(xs, ys) {
    return xs.length >= ys.length && ys.every((y, i) => y === xs[i]);
}

 
class ExtractedConfig {
    constructor() {

         
        this.configNameOfNoInlineConfig = "";

         
        this.env = {};

         
        this.globals = {};

         
        this.ignores = void 0;

         
        this.noInlineConfig = void 0;

         
        this.parser = null;

         
        this.parserOptions = {};

         
        this.plugins = {};

         
        this.processor = null;

         
        this.reportUnusedDisableDirectives = void 0;

         
        this.rules = {};

         
        this.settings = {};
    }

     
    toCompatibleObjectAsConfigFileContent() {
        const {
             
            configNameOfNoInlineConfig: _ignore1,
            processor: _ignore2,
             
            ignores,
            ...config
        } = this;

        config.parser = config.parser && config.parser.filePath;
        config.plugins = Object.keys(config.plugins).filter(Boolean).reverse();
        config.ignorePatterns = ignores ? ignores.patterns : [];

        // Strip the default patterns from `ignorePatterns`.
        if (startsWith(config.ignorePatterns, IgnorePattern.DefaultPatterns)) {
            config.ignorePatterns =
                config.ignorePatterns.slice(IgnorePattern.DefaultPatterns.length);
        }

        return config;
    }
}

export { ExtractedConfig };
