// @napi-rs/canvas/node-canvas.js
'use strict'

const { Readable } = require('node:stream')

const {
  createCanvas: _createCanvas,
  Canvas,
  CanvasElement,
  SVGCanvas,
  GlobalFonts,
  Image,
  ImageData,
  Path2D,
  DOMPoint,
  DOMMatrix,
  DOMRect,
  loadImage,
  PDFDocument,
  SvgExportFlag,
} = require('./index.js')


const { CanvasRenderingContext2D } = require('./js-binding')


const NODE_CANVAS_DEFAULT_QUALITY = 0.75


class PNGStream extends Readable {
  constructor(canvas, options) {
    super()
    this._canvas = canvas
    this._options = options || {}
    this._done = false
  }

  _read() {
    if (this._done) return
    this._done = true
    this._canvas
      .encode('png')
      .then((buf) => {
        this.push(buf)
        this.push(null)
      })
      .catch((err) => {
        this.destroy(err)
      })
  }
}

class JPEGStream extends Readable {
  constructor(canvas, options) {
    super()
    this._canvas = canvas
    const opts = options || {}
    // node-canvas quality: 0-1 (default 0.75), encode() expects 0-100
    this._quality = Math.round((opts.quality != null ? opts.quality : NODE_CANVAS_DEFAULT_QUALITY) * 100)
    this._done = false
  }

  _read() {
    if (this._done) return
    this._done = true
    this._canvas
      .encode('jpeg', this._quality)
      .then((buf) => {
        this.push(buf)
        this.push(null)
      })
      .catch((err) => {
        this.destroy(err)
      })
  }
}

// ---------------------------------------------------------------------------


const MIME_FORMAT_MAP = {
  'image/png': 'png',
  'image/jpeg': 'jpeg',
  'image/webp': 'webp',
  'image/avif': 'avif',
  'image/gif': 'gif',
}

 
function _extractQuality(mime, configOrQuality) {
  if (mime !== 'image/jpeg' && mime !== 'image/webp') return undefined
  if (configOrQuality == null) return NODE_CANVAS_DEFAULT_QUALITY
  if (typeof configOrQuality === 'number') return configOrQuality
  if (typeof configOrQuality === 'object' && configOrQuality.quality != null) {
    return configOrQuality.quality
  }
  return NODE_CANVAS_DEFAULT_QUALITY
}

// ---------------------------------------------------------------------------


const _origToBuffer = CanvasElement.prototype.toBuffer
const _origToDataURL = CanvasElement.prototype.toDataURL

function _compatCreatePNGStream(options) {
  return new PNGStream(this, options)
}

function _compatCreateJPEGStream(options) {
  return new JPEGStream(this, options)
}

 
function _compatToBuffer(mimeOrCallback, configOrQuality) {
  // Callback form: toBuffer(callback) or toBuffer(callback, mime, config)
  if (typeof mimeOrCallback === 'function') {
    const callback = mimeOrCallback
    const mime = typeof configOrQuality === 'string' ? configOrQuality : 'image/png'
    const config = arguments[2]

    if (mime === 'raw') {
      let buf
      try {
        buf = this.data()
      } catch (err) {
        callback(err)
        return
      }
      callback(null, buf)
      return
    }

    const format = MIME_FORMAT_MAP[mime]
    if (!format) {
      callback(new TypeError(`Unsupported MIME type "$${mime}". Supported: $${Object.keys(MIME_FORMAT_MAP).join(', ')}`))
      return
    }
    const q = _extractQuality(mime, config)
    this.encode(format, q != null ? Math.round(q * 100) : undefined).then(
      (buf) => callback(null, buf),
      (err) => callback(err),
    )
    return
  }

  // Sync: no args → PNG
  if (mimeOrCallback === undefined) {
    return _origToBuffer.call(this, 'image/png')
  }

  // Sync: raw pixel data
  if (mimeOrCallback === 'raw') {
    return this.data()
  }

  // Sync: extract quality (0-1) and scale to 0-100 for native toBuffer
  const q = _extractQuality(mimeOrCallback, configOrQuality)
  return _origToBuffer.call(this, mimeOrCallback, q != null ? Math.round(q * 100) : undefined)
}

 
function _compatToDataURL(mimeOrCallback, qualityOrCallback) {
  // toDataURL(callback)
  if (typeof mimeOrCallback === 'function') {
    this.toDataURLAsync('image/png').then(
      (url) => mimeOrCallback(null, url),
      (err) => mimeOrCallback(err),
    )
    return
  }
  // toDataURL(mime, callback)
  if (typeof qualityOrCallback === 'function') {
    this.toDataURLAsync(mimeOrCallback, _extractQuality(mimeOrCallback, undefined)).then(
      (url) => qualityOrCallback(null, url),
      (err) => qualityOrCallback(err),
    )
    return
  }
  // toDataURL(mime, quality, callback)
  const cb = arguments[2]
  if (typeof cb === 'function') {
    // Native toDataURLAsync expects 0-1 (f64) — pass quality through as-is
    const quality = _extractQuality(mimeOrCallback, qualityOrCallback)
    this.toDataURLAsync(mimeOrCallback, quality).then(
      (url) => cb(null, url),
      (err) => cb(err),
    )
    return
  }

  // Sync form: native toDataURL expects 0-1 (f64) — pass quality through as-is
  const quality = _extractQuality(mimeOrCallback, qualityOrCallback)
  return _origToDataURL.call(this, mimeOrCallback, quality)
}

 
function _addCompatMethods(canvas) {
  canvas.createPNGStream = _compatCreatePNGStream
  canvas.createJPEGStream = _compatCreateJPEGStream
  canvas.toBuffer = _compatToBuffer
  canvas.toDataURL = _compatToDataURL
  return canvas
}

// ---------------------------------------------------------------------------


const _registeredFontKeys = []

 
function registerFont(path, fontFace) {
  if (!fontFace || typeof fontFace.family !== 'string') {
    throw new TypeError('registerFont requires a fontFace with a "family" property')
  }
  const key = GlobalFonts.registerFromPath(path, fontFace.family)
  if (!key) {
    throw new Error(`Failed to register font from "$${path}" with family "$${fontFace.family}"`)
  }
  _registeredFontKeys.push(key)
}

 
function deregisterAllFonts() {
  if (_registeredFontKeys.length > 0) {
    GlobalFonts.removeBatch(_registeredFontKeys)
    _registeredFontKeys.length = 0
  }
}

// ---------------------------------------------------------------------------


 
function createCanvas(width, height, type) {
  if (type === 'svg') {
    // SvgExportFlag enum requires a valid variant; NoPrettyXML (0x02) is the
    
    return _createCanvas(width, height, SvgExportFlag.NoPrettyXML)
  }
  if (type === 'pdf') {
    throw new Error(
      'createCanvas with type "pdf" is not supported. Use the PDFDocument class from @napi-rs/canvas directly.',
    )
  }
  if (type != null && type !== 'image') {
    throw new TypeError(`createCanvas: unknown type "$${type}". Supported types: "image", "svg".`)
  }
  return _addCompatMethods(_createCanvas(width, height))
}

 
function createImageData(dataOrWidth, widthOrHeight, height) {
  if (typeof dataOrWidth === 'number') {
    return new ImageData(dataOrWidth, widthOrHeight)
  }
  if (height != null) {
    return new ImageData(dataOrWidth, widthOrHeight, height)
  }
  return new ImageData(dataOrWidth, widthOrHeight)
}

// ---------------------------------------------------------------------------


const CompatCanvas = new Proxy(Canvas, {
  construct(target, args) {
    const canvas = new target(...args)
    if (canvas instanceof CanvasElement) {
      return _addCompatMethods(canvas)
    }
    return canvas
  },
})


module.exports = {
  // Factory functions
  Canvas: CompatCanvas,
  createCanvas,
  createImageData,
  loadImage,
  registerFont,
  deregisterAllFonts,

  // Classes
  Image,
  ImageData,
  CanvasRenderingContext2D,
  Context2d: CanvasRenderingContext2D,
  PNGStream,
  JPEGStream,
  Path2D,

  // Geometry
  DOMPoint,
  DOMMatrix,
  DOMRect,

  // @napi-rs/canvas extras (available but not part of node-canvas)
  GlobalFonts,
  PDFDocument,
  CanvasElement,
  SVGCanvas,
}
